### DISCUSSION TITLE
I2S init() to support multiple files sample_rate/bit_widths?

### DISCUSSION THREAD

**Discussion by zachmoshe** [micropython/micropython#9139] [Libraries & Drivers] [answered]
Hi,
I'm implementing a small ESP32 based device that plays WAV files from a SDCard through a I2S DAC. Since the I2S object allocates a buffer, I want to reuse it between songs. However, every song can potentially be recorded with different sample_rate and bit_width and those are given as arguments in the I2S constructor.

By its name, it looks like `I2S.init()` is a solution but I couldn't find out when it should be used and whether it allocates a new buffer or not. The documentation there doesn't say much, just states that the arguments are similar to the constructor ones. Could anyone clarify? Or provide a best practice of when should I use "init()"?

Should I create an i2s object once in the lifetime of my program with the desired *output* parameters (sample_rate, bit_width) and convert whatever file I read to those? Is there any existing micropython package to help me with resample of WAV samples? Is there any existing implementation for decoding other formats? (mp3, ogg, aac) I'll happily save some space on the SDCard, I'm currently dealing with ~400-500MBs just to keep a few songs...

Thanks!

**Comment by robert-hh** [comment, marked as answer]
The i2s constructor and i2s.init() share all parameters beside the id argument of the constructor. Calling either of them does a  full reconfiguration of the I2S device. If you  call the constructor for an ID that has already been used, the existing object will be re-used and the same will happen as calling i2s.init(). So - there is no difference whether you create the i2s object once and run i2s.init() afterwards or if you call the constructor every time.

I do not know of any implementation for decoding. A python implementation may be too slow.

**Reply by zachmoshe** [reply]
OK, I kind of suspected that.. 
I want to keep a single object running throughout the run so I avoid the "click" sounds I get when I reinitialize the object for every song (I'd guess it's because voltage is applied on the speaker inputs and pushes the membrane to one of the sides, but not 100% sure). 
Just to make sure I'm not missing anything - the only way to do that is either to decide on my "output" parameters and convert all audio files to WAV with the same params, or use a hardware decoder, right?

**Reply by robert-hh** [reply]
If the click happens when you re-create or re-init I2S, them using the same parameters for each song seems to be the only solution. The espressif API does not seem to allow reconfiguration without calling disable() first. I guess that the click is caused by switching the DAC output on and off. It might be possible to keep the output in about the level it has when playing, reducing the voltage swing. But that is analog fiddling.
